mono/metadata/android.c

```diff
@@ -0,0 +1,537 @@
+   solution
+
+   sp_value buffer should be at least PROP_VALUE_MAX+1 bytes long
+*/
+static gint
+_monodroid__system_property_get (const gchar *name, gchar *sp_value, gsize sp_value_len)
+{
+	if (!name)
+		return -1;
+
+	g_assert (sp_value);
+	g_assert (sp_value_len == PROP_VALUE_MAX + 1);
+
+	gchar *cmd = g_strdup_printf ("getprop %s", name);
+	FILE* result = popen (cmd, "r");
+	gint len = (gint) fread (sp_value, 1, sp_value_len, result);
+	fclose (result);
+	sp_value [len] = 0;
+	if (len > 0 && sp_value [len - 1] == '\n') {
+		sp_value [len - 1] = 0;
+		len--;
+	} else {
+		if (len != 0)
+			len = 0;
+		sp_value [0] = 0;
+	}
+
+	g_message ("%s %s: '%s' len: %d", __func__, name, sp_value, len);
+
+	return len;
+}
+#else
+static gint
+_monodroid__system_property_get (const gchar *name, gchar *sp_value, gsize sp_value_len)
+{
+	if (!name)
+		return -1;
+
+	g_assert (sp_value);
+	g_assert (sp_value_len == PROP_VALUE_MAX + 1);
+
+	return __system_property_get (name, sp_value);
+}
+#endif // ANDROID
+
+gint32
+monodroid_get_system_property (const gchar *name, gchar **value)
+{
+	gchar  buf [PROP_VALUE_MAX+1] = { 0, };
+	gint   len;
+	struct BundledProperty *p;
+
+	g_assert (value);
+	*value = NULL;
+
+	len = _monodroid__system_property_get (name, buf, sizeof (buf));
+	if (len > 0) {
+		*value = g_strndup (buf, len);
+		return len;
+	}
+
+	if ((p = lookup_system_property (name))) {
+		*value = g_strndup (p->value, p->value_len);
+		return p->value_len;
+	}
+
+	return -1;
+}
+
+gpointer
+_monodroid_timezone_get_default_id (void)
+{
+	JNIEnv *env = mono_jvm_get_jnienv ();
+	jobject d = (*env)->CallStaticObjectMethod (env, TimeZone_class, TimeZone_getDefault);
+	jstring id = (*env)->CallObjectMethod (env, d, TimeZone_getID);
+	const gchar *mutf8 = (*env)->GetStringUTFChars (env, id, NULL);
+
+	gchar *def_id = g_strdup (mutf8);
+
+	(*env)->ReleaseStringUTFChars (env, id, mutf8);
+	(*env)->DeleteLocalRef (env, id);
+	(*env)->DeleteLocalRef (env, d);
+
+	return def_id;
+}
+
+void
+monodroid_free (gpointer ptr)
+{
+	g_free (ptr);
+}
+
+#if HOST_ANDROID && __arm__
+
+#define BUF_SIZE 512
+
+static gboolean
+find_in_maps (const gchar *str)
+{
+	FILE  *maps;
+	gchar *line;
+	gchar  buf [BUF_SIZE];
+
+	g_assert (str);
+
+	maps = fopen ("/proc/self/maps", "r");
+	if (!maps)
+		return FALSE;
+
+	while ((line = fgets (buf, BUF_SIZE, maps))) {
+		if (strstr (line, str)) {
+			fclose (maps);
+			return TRUE;
+		}
+	}
+
+	fclose (maps);
+	return FALSE;
+}
+
+static gboolean
+detect_houdini ()
+{
+	return find_in_maps ("libhoudini");
+}
+
+#endif // HOST_ANDROID && __arm__
+
+static gboolean
+is_64_bit (void)
+{
+	return SIZEOF_VOID_P == 8;
+}
+
+#define CPU_KIND_UNKNOWN ((guint16)0)
+#define CPU_KIND_ARM     ((guint16)1)
+#define CPU_KIND_ARM64   ((guint16)2)
+#define CPU_KIND_MIPS    ((guint16)3)
+#define CPU_KIND_X86     ((guint16)4)
+#define CPU_KIND_X86_64  ((guint16)5)
+
+static guint16
+get_built_for_cpu (void)
+{
+#if HOST_WIN32
```

Comment:
Suggest you don't really need to check host os, that the next level symbols are unambiguous across any OS/compiler.

i.e.
```
#if _M_AMD64 || _M_X64  || __x86_64__
	return CPU_KIND_X86_64;
#elif _M_IX86 || __i386__
	return CPU_KIND_X86;
#elif _M_ARM || __arm__
	return CPU_KIND_ARM;
#elif __aarch64__
	return CPU_KIND_ARM64;
#elif __mips__
	return CPU_KIND_MIPS;
#else
	return CPU_KIND_UNKNOWN;
#endif
```